# AfterWork Data Science: Data Wrangling with Python Project

## 1. Defining the Question

### a) Specifying the Data Analysis Question

Specify the reasearch question that you'll be answering.

> i.e. Which was the most improved country year 2017 - 2018 as per the GDP ranking?

### b) Defining the Metric for Success

How will you know that your solution will have satisfied your research question?

### c) Understanding the context 

Provide some background information....

### d) Recording the Experimental Design

Describe the steps/approach that you will use to answer the given question.

### e) Data Relevance

How relevant was the provided data?

## 2. Reading the Data

In [1]:
# Importing our libraries
# ---
#
import pandas as pd
import numpy as np

In [2]:
# Load the data below
# --- 
# Dataset url = https://bit.ly/BusBreakdownDataset
# --- 
# 
busbreakdown= pd.read_csv('https://bit.ly/BusBreakdownDataset')

In [3]:
# Checking the first 5 rows of data
# ---
#
busbreakdown.head()

School_Year  Busbreakdown_ID  ... Breakdown_or_Running_Late School_Age_or_PreK
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

In [4]:
# Checking the last 5 rows of data
# ---
#
busbreakdown.tail()

School_Year  ...  School_Age_or_PreK
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[5 rows x 21 columns]

In [5]:
# Sample 10 rows of data
# ---
#
busbreakdown.sample(10)

School_Year  ...  School_Age_or_PreK
163328   2018-2019  ...          School-Age
215348   2017-2018  ...          School-Age
168537   2017-2018  ...          School-Age
261761   2015-2016  ...          School-Age
110041   2017-2018  ...          School-Age
83267    2016-2017  ...               Pre-K
278595   2016-2017  ...          School-Age
266784   2017-2018  ...          School-Age
141220   2017-2018  ...          School-Age
240162   2016-2017  ...          School-Age

[10 rows x 21 columns]

In [6]:
# Checking number of rows and columns
# ---
#  
busbreakdown.shape

(281110, 21)

In [7]:
# Checking datatypes
# ---
# 
busbreakdown.dtypes

School_Year                        object
Busbreakdown_ID                     int64
Run_Type                           object
Bus_No                             object
Route_Number                       object
Reason                             object
Schools_Serviced                   object
Occurred_On                        object
Created_On                         object
Boro                               object
Bus_Company_Name                   object
How_Long_Delayed                   object
Number_Of_Students_On_The_Bus       int64
Has_Contractor_Notified_Schools    object
Has_Contractor_Notified_Parents    object
Have_You_Alerted_OPT               object
Informed_On                        object
Incident_Number                    object
Last_Updated_On                    object
Breakdown_or_Running_Late          object
School_Age_or_PreK                 object
dtype: object

Record your observations below:

*   Observation 1: There are 281,110 records with some columns containing missing data
*   Observation 2: Various columns contain different data i.e route numbers with some values being whole number only and others starting off with a letter followed by 3 numbers



## 3. External Data Source Validation

You can make sure your data matches other credible information. This ensures that the measurements are roughly in line with what they should be and it serves as a check on what other things might be wrong in your dataset.

External validation can often be as simple as checking your data against some other data i.e. country population your dataset with country population on world bank data.

Replace the text found on this text cell with your external data source validation information.

In [8]:
boros = busbreakdown['Boro'].unique()
boros
# NYC has 5 boroughs known as Bronx, Brooklyn, Manhattan, Queens and Staten Island
# There are 5 values under borough column that are not boroughs within the city
# and might be wrong

array(['New Jersey', 'Manhattan', 'Bronx', 'Westchester', 'Brooklyn',
       'Rockland County', 'Nassau County', nan, 'Queens', 'Staten Island',
       'Connecticut', 'All Boroughs'], dtype=object)

In [9]:
busroutes = busbreakdown['Route_Number'].unique()
busroutes
# bus routes seem to be named beginning with the 1st letter of the borough followed 
# by a number. There are various route number values with only numbers or wit
# letters at the end vs the start, indictating some values may be wrong.

array(['J711', 'M351', '3', ..., '012', '29AM', '1409B'], dtype=object)

## 4. Data Preparation

### Performing Data Cleaning

In [10]:
# Checking datatypes and missing entries of all the variables
# ---
# 
busbreakdown.isnull().sum()

School_Year                             0
Busbreakdown_ID                         0
Run_Type                                3
Bus_No                                  9
Route_Number                            7
Reason                                  2
Schools_Serviced                        7
Occurred_On                             0
Created_On                              0
Boro                                13461
Bus_Company_Name                        0
How_Long_Delayed                    35608
Number_Of_Students_On_The_Bus           0
Has_Contractor_Notified_Schools         0
Has_Contractor_Notified_Parents         0
Have_You_Alerted_OPT                    0
Informed_On                             0
Incident_Number                    271627
Last_Updated_On                         0
Breakdown_or_Running_Late               0
School_Age_or_PreK                      0
dtype: int64

We observe the following from our dataset:

*   Observation 1: 271,627 i.e majority of the recorded incidences do not contain incident numbers, this may be important for our analysis later as it could be a key differentiating column 
*   Observation 2: how long delayed and boroughs are also missing entries, that could affect analysis later on



In [11]:
# Standardizing your dataset i.e. variable renaming
# ---
# 
busbreakdown.columns = busbreakdown.columns.str.strip().str.lower()
busbreakdown.head(1)

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age

[1 rows x 21 columns]

We observe the following from our dataset:

*   Observation 1: all columns are now lower case 
*   Observation 2



In [12]:
# Checking how many duplicate rows are there in the data
# ---
# 
sum(busbreakdown.duplicated())  

0

We observe the following from our dataset:

*   Observation 1: There are no duplicate rows in the data



In [13]:
# Checking if any of the columns are all null
# ---
# 
#busbreakdown.isnull().any()
busbreakdown.isnull().sum()

school_year                             0
busbreakdown_id                         0
run_type                                3
bus_no                                  9
route_number                            7
reason                                  2
schools_serviced                        7
occurred_on                             0
created_on                              0
boro                                13461
bus_company_name                        0
how_long_delayed                    35608
number_of_students_on_the_bus           0
has_contractor_notified_schools         0
has_contractor_notified_parents         0
have_you_alerted_opt                    0
informed_on                             0
incident_number                    271627
last_updated_on                         0
breakdown_or_running_late               0
school_age_or_prek                      0
dtype: int64

We observe the following from our dataset:

*   Observation 1: Since there are 281,110 records, we can assume there is no column with all nulls as all null counts are less than the no of records




In [14]:
# Checking if any of the rows are all null
# ---
#busbreakdown.isnull().any()
busbreakdown.shape
busbreakdown_rowna = busbreakdown.dropna(how='all')
busbreakdown_rowna.shape

(281110, 21)

We observe the following from our dataset:

*   Observation 1: dropping all rows where all values are null has not chnaged the df shape showing there are no rows that are all null. majority have entries with only few missing entries in specific collumns

*   Observation 2



In [15]:
# Checking if the "Yes/No" fields contain only these 2 values
# for have_you_alerted_opt variable
# ---
# Hint: Use unique() function
#
busbreakdown['have_you_alerted_opt'].unique()

array(['Yes', 'No'], dtype=object)

We observe the following from our dataset:

*   Observation 1: only contains yes or no values
*   Observation 2: data type is object 



In [16]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_parents variable
# ---
# 
busbreakdown['has_contractor_notified_parents'].unique()

array(['No', 'Yes'], dtype=object)

We observe the following from our dataset:

*   Observation 1: yes, the field contains only 'yes' and 'no' values
*   Observation 2: dtype is still object



In [17]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_schools variable
# ---
#  
busbreakdown['has_contractor_notified_schools'].unique()

array(['Yes', 'No'], dtype=object)

We observe the following from our dataset:

*   Observation 1: Observation 1: yes, the field contains only 'yes' and 'no' values
*   Observation 2: dtype is still object



In [18]:
# Checking unique values in break_down_or_running_late variable to ensure there is no duplication
# ---
#
busbreakdown['breakdown_or_running_late'].unique()

array(['Running Late', 'Breakdown'], dtype=object)

In [19]:
# Checking unique values in school_age_or_prek variable
# ---
# 
busbreakdown['school_age_or_prek'].unique()

array(['School-Age', 'Pre-K'], dtype=object)

In [20]:
# Checking unique values in school_year variable
# ---
#  
busbreakdown['school_year'].unique()

array(['2015-2016', '2017-2018', '2018-2019', '2016-2017', '2019-2020'],
      dtype=object)

In [21]:
# Checking unique values in reason variable
# ---
#  
busbreakdown['reason'].unique()

array(['Heavy Traffic', 'Flat Tire', 'Other', 'Won`t Start',
       'Mechanical Problem', 'Problem Run', 'Accident',
       'Late return from Field Trip', 'Delayed by School',
       'Weather Conditions', nan], dtype=object)

In [22]:
# Checking unique values in run_type variable
# ---
#  
busbreakdown['run_type'].unique()

array(['Special Ed AM Run', 'Pre-K/EI', 'General Ed AM Run',
       'General Ed Field Trip', 'Special Ed PM Run', 'General Ed PM Run',
       'Project Read PM Run', 'Special Ed Field Trip',
       'Project Read AM Run', 'Project Read Field Trip', nan],
      dtype=object)

In [23]:
# Checking unique values in boro variable
# ---
#   
busbreakdown['boro'].unique()

array(['New Jersey', 'Manhattan', 'Bronx', 'Westchester', 'Brooklyn',
       'Rockland County', 'Nassau County', nan, 'Queens', 'Staten Island',
       'Connecticut', 'All Boroughs'], dtype=object)

### Overall Data Cleaning Observations
**Missing Values**

- There are a large number of missing values in the fields "How_Long_Delayed" which is important to our analysis. 
- There is an extremely large number of missing values in the "Incident_number" field but this is not incidental to our analysis and cannot be filled in without additional information.

**Error in values**

- "How_Long_Delayed" contains string values such as "MINS" or "mins" and a range of values, which needs to be changed to single integer value for our analysis.

**Error in Datatypes**

- "How_Long_Delayed" is a string datatype, should be converted to integer type.

**Error in field names**
- The column name "Boro" should be renamed to "Borough".


### Next Steps: Data Cleaning Steps

**Error in values**

- Extract the first integer value (lowest delay time) in the column "How_Long_Delayed"


**Missing Values**

- Impute the missing values in the field "How_Long_Delayed" with the mean value.


**Error in Datatypes**

- Convert "How_Long_Delayed" to int datatype.



**Error in field names**

- Rename the column "Boro" to "Borough".

In [24]:
# Lets first start by creating a copy of our dataframe 
# df_clean = df.copy(). We will use this copy as our cleaning copy.
# ---
#
busbreakdown_clean = busbreakdown.copy()
busbreakdown_clean.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

In [25]:
# Then extracting the lowest delay time in the column how_long_delayed from the string
# 
busbreakdown_clean['how_long_delayed'] = busbreakdown_clean['how_long_delayed'].str.extract('(\d+)')
busbreakdown_clean['how_long_delayed'].head()

0    NaN
1     20
2     15
3     15
4    NaN
Name: how_long_delayed, dtype: object

We impute the null values in 'how_long_delayed' column with mean of the column. This will take a couple of steps...

In [26]:
# We first convert our how_long_delayed to float type to allow for imputation
# ---
# 
#busbreakdown_clean.dtypes
#busbreakdown_clean['how_long_delayed'] = busbreakdown_clean['how_long_delayed'].astype(float)
busbreakdown_clean['how_long_delayed'] = pd.to_numeric(busbreakdown_clean['how_long_delayed'],errors = 'coerce')
busbreakdown_clean.dtypes

school_year                         object
busbreakdown_id                      int64
run_type                            object
bus_no                              object
route_number                        object
reason                              object
schools_serviced                    object
occurred_on                         object
created_on                          object
boro                                object
bus_company_name                    object
how_long_delayed                   float64
number_of_students_on_the_bus        int64
has_contractor_notified_schools     object
has_contractor_notified_parents     object
have_you_alerted_opt                object
informed_on                         object
incident_number                     object
last_updated_on                     object
breakdown_or_running_late           object
school_age_or_prek                  object
dtype: object

In [27]:
# Then later perform our mean imputation 
# ---
# 
#busbreakdown_clean['how_long_delayed'].mean()
busbreakdown_clean['how_long_delayed'].fillna(busbreakdown_clean['how_long_delayed'].mean(),inplace= True)

In [30]:
# Then convert back our how_long_delayed column to integer datatype
# ---
# 
busbreakdown_clean['how_long_delayed'] = busbreakdown_clean['how_long_delayed'].astype(int)
busbreakdown_clean.dtypes

school_year                        object
busbreakdown_id                     int64
run_type                           object
bus_no                             object
route_number                       object
reason                             object
schools_serviced                   object
occurred_on                        object
created_on                         object
boro                               object
bus_company_name                   object
how_long_delayed                    int64
number_of_students_on_the_bus       int64
has_contractor_notified_schools    object
has_contractor_notified_parents    object
have_you_alerted_opt               object
informed_on                        object
incident_number                    object
last_updated_on                    object
breakdown_or_running_late          object
school_age_or_prek                 object
dtype: object

In [29]:
# Then check for nulls in the column
# ---
# 
busbreakdown_clean['how_long_delayed'].isnull().sum()

0

In [38]:
# Rename Boro column to Borough
# ---
# 
busbreakdown_clean.rename(columns={"boro": "borough"}, inplace = True)
busbreakdown_clean.head(1)

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age

[1 rows x 21 columns]

In [39]:
# Lastly we convert all values in our colums to lower case
# for ease of reading
# ---
# 
busbreakdown_clean.columns = busbreakdown_clean.columns.str.lower()
busbreakdown_clean.head(1)

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age

[1 rows x 21 columns]

In [40]:
# Check the first 5 record the cleaned dataset
# ---
#
busbreakdown_clean.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]


## 5. Solution Implementation

Here we investigate the questions that would help craft our recommendations.

### 5.a) Questions

In [46]:
# 1. Which bus companies that had the highest breakdowns?
# ---
#
breakdowns = busbreakdown_clean.groupby(['bus_company_name']).count()

# Sort to get bus company with highest breakdowns
# ---
# YOUR CODE GOES BELOW
#
breakdowns.sort_values(by='busbreakdown_id',ascending=0)

school_year  ...  school_age_or_prek
bus_company_name                                 ...                    
G.V.C., LTD.                              19394  ...               19394
LEESEL TRANSPORTATION CORP (B2192)        17200  ...               17200
RELIANT TRANS, INC. (B232                 13741  ...               13741
PIONEER TRANSPORTATION CO                 12017  ...               12017
BORO TRANSIT, INC.                        11953  ...               11953
...                                         ...  ...                 ...
SMART PICK INC                                1  ...                   1
R & C TRANSIT, INC. (B2321)                   1  ...                   1
L&M Bus Corp.                                 1  ...                   1
FORTUNA BUS COMPANY                           1  ...                   1
phillip bus service                           1  ...                   1

[117 rows x 20 columns]

In [49]:
# 2. What were the top 3 reasons for bus delays?
# ---
#
bus_delay = busbreakdown_clean.groupby(['reason']).count()[['how_long_delayed']].reset_index()

# Sort to get the most frequent reason
# ---
# YOUR CODE GOES BELOW
#
bus_delay.sort_values(by='how_long_delayed', ascending = 0)

reason  how_long_delayed
3                Heavy Traffic            173221
6                        Other             37579
5           Mechanical Problem             28162
9                  Won`t Start             12283
2                    Flat Tire              8411
8           Weather Conditions              6932
4  Late return from Field Trip              5704
7                  Problem Run              4068
0                     Accident              2526
1            Delayed by School              2222

In [53]:
# 3. How many students were in the buses when they broke down? 
# ---
#
studentsonbus = busbreakdown_clean.groupby(['number_of_students_on_the_bus']).count()[['busbreakdown_id']].reset_index()
studentsonbus['number_of_students_on_the_bus'].sum() 

310996

In [59]:
# 4. Which were most frequent reasons for bus breakdowns?
# ---
#
breakdown_reasons = busbreakdown_clean[busbreakdown_clean.breakdown_or_running_late == 'breakdown'].groupby(['reason']).count()

# Sort to get most frequent reasons
# ---
# YOUR CODE GOES BELOW
#
breakdown_reasons.head(2)
#breakdown_reasons.sort_values(by='how_long_delayed', ascending = 0)

Empty DataFrame
Columns: [school_year, busbreakdown_id, run_type, bus_no, route_number, schools_serviced, occurred_on, created_on, borough, bus_company_name, how_long_delayed, number_of_students_on_the_bus, has_contractor_notified_schools, has_contractor_notified_parents, have_you_alerted_opt, informed_on, incident_number, last_updated_on, breakdown_or_running_late, school_age_or_prek]
Index: []

In [80]:
# 5. What were the most frequent reasons for the bus running late? 
# ---
#
reasons = busbreakdown_clean.groupby(['reason','how_long_delayed']).count()

# Get the records with running late reasons and sort to get most frequent reasons
# ---
# YOUR CODE GOES BELOW
#
#reasons['breakdown_or_running_late']
#reasons.sort_values(by='', ascending = 0)
running_late = busbreakdown_clean[busbreakdown_clean.breakdown_or_running_late.isin(['Running Late'])].count()
#running_late.sort_values(by='reason', ascending = 0)
running_late.head()

school_year        250053
busbreakdown_id    250053
run_type           250051
bus_no             250044
route_number       250046
dtype: int64

In [88]:
# 6. What was the average delay time for each reason type?
# ---
# 
avg_delay = busbreakdown_clean.groupby('reason').mean().reset_index()

# Get the records with reasons and how long on average a delay took then sort 
# ---
# YOUR CODE GOES BELOW
#
avg_delay.sort_values(by='how_long_delayed', ascending = 0)

reason  ...  number_of_students_on_the_bus
4  Late return from Field Trip  ...                       4.396739
0                     Accident  ...                       6.305622
7                  Problem Run  ...                       6.407080
5           Mechanical Problem  ...                       1.552376
9                  Won`t Start  ...                       0.920215
2                    Flat Tire  ...                       1.720723
8           Weather Conditions  ...                       2.378823
6                        Other  ...                       2.737965
3                Heavy Traffic  ...                       3.926447
1            Delayed by School  ...                       5.882088

[10 rows x 4 columns]

### 5.b) Recommendations

From the above analysis, below are our recommendations:

1.   heavy traffic seemed to be a major reason for bus delays, recommend to use routes with lower traffic or set pick up times to off peak hours
2.   more evenly distribute the schools services by 1 bus compay to reduce lateness or breakdowns





## 6. Challenging your Solution

During this step, we review our solution and implement  approaches that could potentially provide a better outcome. In our case, we could propose the following question that wasn't answered in our solution because it couldn't have greatly contributed to our recommendation.

In [84]:
# Which boroughs experienced the most breakdowns?
# ---  
#
breakdowns = busbreakdown_clean.groupby(['borough']).count()

# Sort to get bus company with highest breakdowns
# ---
# YOUR CODE GOES BELOW
#
breakdowns.sort_values(by='bus_company_name',ascending=0)

school_year  ...  school_age_or_prek
borough                       ...                    
Bronx                  70952  ...               70952
Manhattan              68177  ...               68177
Brooklyn               62239  ...               62239
Queens                 38060  ...               38060
Staten Island          13907  ...               13907
Westchester             7093  ...                7093
Nassau County           4124  ...                4124
New Jersey              1583  ...                1583
Rockland County          875  ...                 875
All Boroughs             429  ...                 429
Connecticut              210  ...                 210

[11 rows x 20 columns]

Our observations:

1.   Bronx borough has the highest no of breakdowns, conneticut has the lowest
2.  


How does this observation tie to our solution? ANSWER:  this observation shows there could be factors at play in bronx affecting buses that may not be present in those ares with lower breakdowns



## 7. Follow up questions

During this step, you rethink and propose other ways that you can improve your solution.

### a). Did we have the right data?

the data was relevant, however we needed other data to improve the solution

### b). Do we need other data to answer our question?

You can look into the questions you brainstormed that you weren't taken into account during analysis due to a lack of data. Were those questions important to have been left out  of your analysis?

Answer: we were missing key data such as incident IDs, to identify if these cases were unique or logged on the system at all for a more complete view of the data

### c). Did we have the right question?

Were there any other questions that we needed to have answered?

Answer: yes,
*   what routes had the highest number of buses running late or delayed
*   opt alerted routes vs parents alerted
*   number of schools services compared to breakdowns

